### PR TITLE
TE FP8 integration

### PR DESCRIPTION
**Pull request by SanjuCSudhakaran**
# What does this PR do?

Refactoring `transformer_engine` FP8 support in optimum-habana. This removes model specific changes and makes the code more model agnostic.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: examples/language-modeling/README.md [hunk 1]
```diff
@@ -574,6 +574,46 @@ python3 ../gaudi_spawn.py --use_deepspeed  --world_size 8  run_lora_clm.py \
   --flash_attention_causal_mask True
 ```
 
+- Multi-card finetuning of Llama2-70B with DeepSpeed ZeRO-3 optimization, LoRA and FP8 precision:
```
Comment (vivekgoe) [inline examples/language-modeling/README.md:577]: Maybe we can just modify existing command for FP8 precision (instead of adding a new one).

File: optimum/habana/transformers/training_args.py [hunk 2]
```diff
@@ -291,14 +291,6 @@ class GaudiTrainingArguments(TrainingArguments):
         metadata={"help": "Whether to use fp8 for training."},
     )
 
-    fp8_recipe_format: Optional[str] = field(
```
Comment (vivekgoe) [inline optimum/habana/transformers/training_args.py:294]: @libinta please note we deleted this because we do not want to expose low level numeric selection to user as of now. In future we can revisit this.

**Comment by vivekgoe** [comment]
@regisss @libinta please help review this PR. This is a cleaned up version of TE integration (tested with Llama-7B, Llama-70B, Mixtral-7B fine-tuning tasks).  Changes were done mainly for following reasons,
-  Align to newer PEFT version (0.10.0) which helps reducing custom changes to support LoRA layers in fp8,
-  Add generic fp8 support for models using activation_checkpointing, 
-  Align TE code to accelerator repo to extent possible. Limit fp8 configuration interface exposed to user (allow only well tested configuration for now)

**Comment by vivekgoe** [comment]
Related PR with tests https://github.com/huggingface/optimum-habana/pull/1106

File: examples/language-modeling/README.md [hunk 3]
```diff
@@ -574,6 +574,46 @@ python3 ../gaudi_spawn.py --use_deepspeed  --world_size 8  run_lora_clm.py \
...
+
+  > The following command requires Habana DeepSpeed 1.13.0 or later.
+
+```bash
+PT_HPU_MAX_COMPOUND_OP_SIZE=10 \
```
Comment (libinta) [inline examples/language-modeling/README.md:582]: why do we still need PT_HPU_MAX_COMPOUND_OP_SIZE=10 ?

Comment (SanjuCSudhakaran) [reply examples/language-modeling/README.md:582]: There is a spike in the loss curve when the flag is removed, which results in a slight drop in overall accuracy for Llama2-70B.

So for now we want to keep this flag, while we investigate and fix this issue.

File: examples/language-modeling/README.md (hunk 1 above)
Comment (regisss) [reply examples/language-modeling/README.md:577]: I agree with Vivek

Comment (SanjuCSudhakaran) [reply examples/language-modeling/README.md:538]: Done. Updated the readme.

**Review by vivekgoe — APPROVED** [review]

**Review by libinta — APPROVED** [review]

**Review by regisss — APPROVED** [review]
